File: Graphs/802. Find Eventual Safe States/solution.py
class Solution:
    def dfs(self, node, visited, pathvisited, adj,safenodes):
        visited[node] = 1
        pathvisited[node] = 1
        for i in adj[node]:
            if visited[i] == 0:
                if self.dfs(i,visited,pathvisited, adj,safenodes):
                    return True
            elif pathvisited[i] == 1:
                return True
        pathvisited[node] = 0
        safenodes[node] = 1
        return False
    def eventualSafeNodes(self, graph: List[List[int]]) -> List[int]:
        V = len(graph)
        visited = [0 for i in range(V)]
        pathvisited = [0 for i in range(V)]
        safenodes = [0 for i in range(V)]
        ls = []
        for i in range(V):
            if not visited[i]:
                self.dfs(i, visited,pathvisited,graph,safenodes)
        for i in range(V):
            if safenodes[i] == 1:
                ls.append(i)
        return ls

File: Graphs/802. Find Eventual Safe States/solution.cpp
class Solution {
public:
bool dfs(int node, vector<int> &visited, vector<int> &pathvisited, vector<vector<int>> adj,vector<int> &safenodes){
            visited[node] = 1;
            pathvisited[node] = 1;
            for(auto i: adj[node]){
                if(!visited[i]){
                    if(dfs(i,visited,pathvisited, adj,safenodes)){
                        return true;
                    }
                }
                else if(pathvisited[i]){
                    return true;
                }
            }
            pathvisited[node] = 0;
            safenodes[node] = 1;
            return false;
    }
    vector<int> eventualSafeNodes(vector<vector<int>>& graph) {
        int V;
        V = graph.size();
        vector<int> visited(V,0);
        vector<int> pathvisited(V,0);
        vector<int> safenodes(V,0);
        vector<int> ans;
        for(int i = 0; i < V;i++){
            if(!visited[i]){
                dfs(i, visited,pathvisited,graph,safenodes);
            }
        }
        for(int i = 0; i < V;i++){
            if(safenodes[i] == 1){
                ans.push_back(i);
            }
        }
        return ans;
        
    }
        
};
